### PR TITLE
Implement capability to use a validator config file

### PR DESCRIPTION
**Pull request by agunde406**
This commit adds a ValidatorConfig object, an example
validator config file, and updates the validator cli to
check for a config file and combine it with any command
line options set.

Important to note the public-uri is no longer "required" to
be set on the command line, but if it is not set by either
the command line or the config file the validator will
shutdown.

Signed-off-by: Andrea Gunderson <andreax.gunderson@intel.com>

### CODE REVIEW

File: validator/packaging/validator.toml.example [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+#
+
+# This file should exist in the defined config directory and allows
+# validators to be configured without the need for command line options.
+# It is important to note that public_uri must be set by either the config or
+# the command line.
+
+# The following is a possible example.
+
+network_endpoint = "tcp://127.0.0.1:8800"
```
Comment (peterschwarz) [inline validator/packaging/validator.toml.example]: Each of these settings should be documented

File: validator/sawtooth_validator/config/validator.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        if config.component_endpoint is not None:
+            component_endpoint = config.component_endpoint
+        if config.public_uri is not None:
+            public_uri = config.public_uri
+        if config.peering is not None:
+            peering = config.peering
+        if config.join is not None:
+            join = config.join
+        if config.peers is not None:
+            peers = config.peers
```
Comment (aludvik) [inline validator/sawtooth_validator/config/validator.py]: This might be cleaner in a loop. Might be good to make the list of options part of ValidatorConfig too. Something like:
```
merged = ValidatorConfig()
for config in reversed(configs):
  for field in config.list_fields():
    if config.getattr(field):
      merged.setattr(field, config.getattr(field))
```

File: validator/packaging/validator.toml.example (hunk 1 above)
Comment (aludvik) [inline validator/packaging/validator.toml.example]: Aren't they already documented in the validator CLI?. Can we instead reference the cli, explain how to go between the two, and remove the comments from this file so there isn't duplicate documentation?  Eg., "convert remove leading '--' and convert remaining '-' to '_'"

File: validator/sawtooth_validator/config/validator.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import toml
+
+from sawtooth_validator.exceptions import LocalConfigurationError
+from sawtooth_validator.config.path import get_default_path_config
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def load_toml_validator_config(filename):
```
Comment (aludvik) [inline validator/sawtooth_validator/config/validator.py]: Would it be clearer if the top level methods were instead class methods of  ValidatorConfig?

```
load_toml_validator_config -> ValidatorConfig.from_file()
merge_path_config -> ValidatorConfig.merge()
+ ValidatorConfig.from_opts()
```

If the default behaviour of from_file() is then to load from default_path_config.config_dir, then you have:
```
def load_validator_config(opts, config_dir=None):
  opt_conf = ValidatorConfig.from_opts(opts)
  file_conf = ValidatorConfig.from_file(config_dir)
  return ValidatorConfig.merge(opt_conf, file_conf)
```

File: validator/packaging/validator.toml.example (hunk 1 above)
Comment (peterschwarz) [inline validator/packaging/validator.toml.example]: Was thinking about this file in this context: https://github.com/hyperledger/sawtooth-core/blob/8f1ea3a22f0fcf1ccbe994c05c3b4469edaab026/validator/packaging/path.toml.example

File: validator/sawtooth_validator/config/validator.py (hunk 3 above)
Comment (agunde406) [inline validator/sawtooth_validator/config/validator.py]: The format of this file is based off of path.py and the PathConfig object.

Comment (vaporos) [inline validator/sawtooth_validator/config/validator.py]: IMO, the current method is better, as it keeps the data separate from the load mechanism.

from_file() with no arguments is singleton-like, which is extremely poor in that it provides a methodology to take control away from the top-level application -- creating technical debt to be incurred later.  Better to force passing the correct path as determined by the top-level application logic.

Since this module need not know about application-level logic, from_opts() functionality is best left elsewhere.

If putting merge in the ValidatorConfig class, one could see it acting like set's union() and acting on self and others but still returning a new config.

File: validator/sawtooth_validator/server/cli.py [hunk 4]
```diff
@@ -194,17 +200,22 @@ def main(args=sys.argv[1:]):
                            human_readable_name='Log'):
         init_errors = True
 
+    if validator_config.public_uri is None:
```
Comment (vaporos) [inline validator/sawtooth_validator/server/cli.py]: Let's get rid of the need for specifying public_uri in the common case -- there is no reason we have to specify this unless we are binding to multiple interfaces via 0.0.0.0 or behind NAT.  If network_endpoint is set to tcp://0.0.0.0:port and public_uri is not been set, then we should exit.  In all other cases, public_uri should default to network_endpoint.

File: validator/sawtooth_validator/config/validator.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+         network_endpoint=network_endpoint,
+         component_endpoint=component_endpoint,
+         public_uri=public_uri,
+         peering=peering,
+         join=join,
+         peers=peers
+    )
+
+
+def load_validator_config(opts, config_dir=None):
```
Comment (vaporos) [inline validator/sawtooth_validator/config/validator.py]: If I read this correctly, this relies on opts (which came from argparse) having the same fields as ValidatorConfig.

A better method would be to explicitly create a ValidatorConfig() explicitly mapping opts fields to a ValidatorConfig and then merging together the two config objects.  This will produce a more obvious error should one of them change and not the other.

ValidatorConfig(network_endpoint=opts.network_endpoint, ...)

I would also move this function to cli.py since nothing else should know about argparse options.

File: validator/packaging/validator.toml.example [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#
```
Comment (vaporos) [inline validator/packaging/validator.toml.example]: Maybe we should have a unit test which makes sure the values here reflect the actual default values to prevent them from drifting auto-of-sync.

The defaults are probably scattered around various modules currently; we should probably move them to config/validator.py (load_default_validator_config?) and merge them in as well.

File: validator/sawtooth_validator/server/cli.py (hunk 4 above)
Comment (jsmitchell) [inline validator/sawtooth_validator/server/cli.py]: This also needs to handle the case of network_endpoint specifying interface names (e.g. tcp://eth0:8800).

File: validator/packaging/validator.toml.example (hunk 6 above)
Comment (agunde406) [inline validator/packaging/validator.toml.example]: Unit test Added and load_default_validator_config added to config/validator.

File: validator/sawtooth_validator/config/validator.py (hunk 2 above)
Comment (agunde406) [inline validator/sawtooth_validator/config/validator.py]: I disagree with the above, this makes it harder to see what is part of the multiple configs. ValdiatorConfig already has a way to get a dictionary of the fields.

**Comment by agunde406** [comment]
Updated to include the above requested changes, except for changing the --join option as a better name could not be decided on.

File: validator/tests/unit3/test_config/tests.py [hunk 7]
```diff
@@ -135,3 +137,93 @@ def test_path_config_invalid_setting_in_file(self):
...
             os.environ.update(orig_environ)
             shutil.rmtree(directory)
+
+
+class TestValidatorConfig(unittest.TestCase):
+
+    def test_validator_config_defaults(self):
+        """Tests the default validator configuration when no other configs.
+        The defaults should be as follows:
+            - network_endpoint = "tcp://127.0.0.1:8800"
```
Comment (jsmitchell) [inline validator/tests/unit3/test_config/tests.py]: Still have the prior names here.

File: validator/sawtooth_validator/server/cli.py [hunk 8]
```diff
@@ -137,6 +138,26 @@ def _split_comma_append_args(arg_list):
...
+    default_validator_config = load_default_validator_config()
+    conf_file = os.path.join(config_dir, 'validator.toml')
+
+    toml_config = load_toml_validator_config(conf_file)
+
+    return merge_validator_config(
+        configs=[first_config, toml_config, default_validator_config])
+
+
+def create_valdiator_config(opts):
```
Comment (jsmitchell) [inline validator/sawtooth_validator/server/cli.py]: s/valdiator/validator/

File: validator/sawtooth_validator/server/cli.py [hunk 9]
```diff
@@ -157,6 +178,14 @@ def main(args=sys.argv[1:]):
         LOGGER.error(str(local_config_err))
         sys.exit(1)
 
+    try:
+        opts_config = create_valdiator_config(opts)
```
Comment (jsmitchell) [inline validator/sawtooth_validator/server/cli.py]: valdiator

File: validator/sawtooth_validator/server/cli.py [hunk 10]
```diff
@@ -137,6 +138,26 @@ def _split_comma_append_args(arg_list):
...
+
+    toml_config = load_toml_validator_config(conf_file)
+
+    return merge_validator_config(
+        configs=[first_config, toml_config, default_validator_config])
+
+
+def create_valdiator_config(opts):
+    return ValidatorConfig(
+        bind_network=opts.network_endpoint,
```
Comment (jsmitchell) [inline validator/sawtooth_validator/server/cli.py]: Are you going to change the command line args, or will that be a separate story?

Comment (agunde406) [inline validator/sawtooth_validator/server/cli.py]: I am working on it now, but it will be in a separate PR because of the amount of docker file changes necessary.

File: validator/packaging/validator.toml.example [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+# This file should exist in the defined config directory and allows
+# validators to be configured without the need for command line options.
+
+# The following is a possible example.
+
+# Bind is used to set the network and component endpoints. It should be a list
+# of strings in the format "option:endpoint", where the options are currently
+# network and component.
+    bind = [
```
Comment (vaporos) [inline validator/packaging/validator.toml.example]: Should remove the indentation on these settings.
